client/src/worker_registry/mod.rs

```diff
@@ -0,0 +1,275 @@
+    /// The namespace for the WFTs that it can process.
+    fn namespace(&self) -> &str;
+    /// The task queue this provider listens to.
+    fn task_queue(&self) -> &str;
+    /// Try to reserve a slot on this worker.
+    fn try_reserve_wft_slot(&self) -> Option<Box<dyn Slot + Send>>;
+}
+
+/// This trait represents a slot reserved for processing a WFT by a worker.
+#[cfg_attr(test, mockall::automock)]
+pub trait Slot {
+    /// Consumes this slot by dispatching a WFT to its worker. This can only be called once.
+    fn schedule_wft(
+        self: Box<Self>,
+        task: PollWorkflowTaskQueueResponse,
+    ) -> Result<(), anyhow::Error>;
+}
+
+/// This trait enables local workers to made themselves visible to a shared client instance.
+/// There can only be one worker registered per namespace+queue_name+client, others will get ignored.
+#[cfg_attr(test, mockall::automock)]
+pub trait WorkerRegistry {
+    /// Register a local worker that can provide WFT processing slots.
+    fn register(&self, provider: Box<dyn SlotProvider + Send + Sync>) -> Option<WorkerKey>;
+    /// Unregister a provider, typically when its worker starts shutdown.
+    fn unregister(&self, id: WorkerKey);
+}
+
+#[derive(PartialEq, Eq, Hash, Debug, Clone)]
+struct SlotKey {
+    namespace: String,
+    task_queue: String,
+}
+
+impl SlotKey {
+    fn new(namespace: String, task_queue: String) -> SlotKey {
+        SlotKey {
+            namespace,
+            task_queue,
+        }
+    }
+}
+
+/// Implements a [WorkerRegistry] and provides a convenient method
+/// to find compatible slots within the collection.
+/// This is an inner class to hide the mutex.
+#[derive(Default, Debug)]
+struct SlotManagerImpl {
+    /// Maps keys, i.e., namespace#task_queue, to provider.
+    providers: HashMap<SlotKey, Box<dyn SlotProvider + Send + Sync>>,
+    /// Maps ids to keys in `providers`.
+    index: SlotMap<WorkerKey, SlotKey>,
+}
+
+impl SlotManagerImpl {
+    /// Factory method.
+    fn new() -> Self {
+        Self {
+            index: Default::default(),
+            providers: Default::default(),
+        }
+    }
+
+    fn try_reserve_wft_slot(
+        &self,
+        namespace: String,
+        task_queue: String,
+    ) -> Option<Box<dyn Slot + Send>> {
+        let key = SlotKey::new(namespace, task_queue);
+        if let Some(p) = self.providers.get(&key) {
+            if let Some(slot) = p.try_reserve_wft_slot() {
+                return Some(slot);
+            }
+        }
+        None
+    }
+
+    fn register(&mut self, provider: Box<dyn SlotProvider + Send + Sync>) -> Option<WorkerKey> {
+        let key = SlotKey::new(
+            provider.namespace().to_string(),
+            provider.task_queue().to_string(),
+        );
+        if let Vacant(p) = self.providers.entry(key.clone()) {
+            p.insert(provider);
+            Some(self.index.insert(key))
+        } else {
+            warn!("Ignoring registration for worker in bucket {key:?}.");
```

Comment:
"Bucket" isn't going to mean much for a user seeing the warning.
```suggestion
            warn!("Ignoring registration for worker: {key:?}.");
```

Reply:
Done